pinot-broker/src/main/java/org/apache/pinot/broker/routing/RoutingManager.java

```diff
@@ -458,6 +461,35 @@ public TimeBoundaryInfo getTimeBoundaryInfo(String offlineTableName) {
     return timeBoundaryManager != null ? timeBoundaryManager.getTimeBoundaryInfo() : null;
   }
 
+  /**
+   * Fetches segments which replicas are all in ERROR state, put the result to noReplicaSegments set.
+   */
+  public void fetchSegmentsInErrorState(ExternalView externalView, Set<String> noReplicaSegments) {
```

Comment:
(nit): Should we say "Fetches segments which are in error state on all of of their replicas"? IIUC, this is what the function is doing

Reply:
Suggest changing the arg name to segmentsWithNoReplicas

Reply:
Currently this is being called when we build the routing table. Should we also call it upon every external view update/change?

Reply:
Suggest rename method to `findUnavailableSegments` In the comments you can mention that these are only ERROR state segments. Realtime consuming segments that may be unavailable are not treated here but are accounted for in the staleness. If someone gets an idea and starts to add realtime unavailable segments here, we have a problem, right?

Reply:
Also we should rebuild our internal state of unavailable segments upon every external view change.

Reply:
Correct. The logic is in `findUnavailableSegments()` in Line 327.